### PR TITLE
Sidebar lazy load.

### PR DESCRIPTION
**Pull request by baiqiushi**
## Title ##
Top 50 Hashtags and Most Recent 10 Sample Tweets requests will only be sent when necessary (lazily).

## Overview ##
 - In `common/services.js` wrap HashTag request and Sample Tweets request to handle* functions and provide a new interface `querySidebar` inside cloudberry service.
 - In `sidebar/controllers.js` add a new `SidebarCtrl` controller who declares a function `showOrHideSidebar` to handle `open/close sidebar` event and a function `showTab` to handle `change tab` event and they call `querySidebar` accordingly.
 - In `index.scala.html` call `showTab` when tab changed and call `showOrHideSidebar` when sidebar is opened or closed.

### CODE REVIEW

**Comment by chenlica** [comment]
Excellent progress!  It will be good to give an overview of these changes.

@waans11 : can you review it?

**Comment by waans11** [comment]
I checked the code. I may need to go through the changes with him F2F.

**Comment by ilvry** [comment]
I checked the code and tested locally, and I think its good. 
I like the idea of using the same hashmap/sample-tweet Json request for all maps.

File: examples/twittermap/web/app/views/twittermap/index.scala.html [hunk 1]
```diff
@@ -31,13 +31,13 @@
...
       <ul class="nav nav-tabs tabs-left">
-        <li role="presentation" ng-click="click = 1"><a href="#hashtag" data-toggle="tab"><i class="fa fa-hashtag fa-2x" aria-hidden="true"></i></a></li>
-        <li role="presentation" ng-click="click = 1"><a href="#tweet" data-toggle="tab"><i class="fa fa-twitter fa-2x" aria-hidden="true"></i></a></li>
-        <li role="presentation" ng-click="click = 1" class="active"><a href="#about" data-toggle="tab"><i class="fa fa-info-circle fa-2x"></i></a></li>
-        <li role="presentation" ng-click="click = click*-1"><i ng-class="click ===-1 ? 'fa fa-chevron-left fa-2x': 'fa fa-chevron-right fa-2x'"></i></li>
+        <li role="presentation" ng-click="click = 1; showTab('hashtag')"><a href="#hashtag" data-toggle="tab"><i class="fa fa-hashtag fa-2x" aria-hidden="true"></i></a></li>
+        <li role="presentation" ng-click="click = 1; showTab('tweet')"><a href="#tweet" data-toggle="tab"><i class="fa fa-twitter fa-2x" aria-hidden="true"></i></a></li>
+        <li role="presentation" ng-click="click = 1; showTab('about')" class="active"><a href="#about" data-toggle="tab"><i class="fa fa-info-circle fa-2x"></i></a></li>
```
Comment (baiqiushi) [inline examples/twittermap/web/app/views/twittermap/index.scala.html]: Remove this call of showTab('about');

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 2]
```diff
@@ -252,6 +253,74 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
...
+            id: "sample",
+            category: "sample"
+          }
+        }
+      };
+    }
+
+    function handleSampleTweetsRequest(parameters) {
+
+      if (parameters.isSampleTweetsOpen && parameters.maptype === "countmap") {
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/common/services.js]: Add a comment here: why only the countmap needs to be checked.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 3]
```diff
@@ -516,46 +509,25 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
...
             ws.send(pointsTimeJson);
-            ws.send(pinHashTagJson);
             break;
           
           default:
             // unrecognized map type
             break;
         }
+
+        cloudberryService.querySidebar(parameters);
```
[the remark this answers is not in the document]
Comment (baiqiushi) [reply examples/twittermap/web/public/javascripts/common/services.js]: The former one will not work.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 4]
```diff
@@ -1,4 +1,106 @@
 angular.module('cloudberry.sidebar', ['cloudberry.common'])
+  .controller("SidebarCtrl", function($scope, cloudberry) {
+    cloudberry.parameters.isSampleTweetsOpen = false;
+    cloudberry.parameters.isHashTagOpen = false;
+
+    $scope.currentTab = "aboutTab";
+
+    // Remember current hash tag results corresponding query parameters
+    $scope.hashtagCurrentParameters =  {
+      keywords: Array.from(cloudberry.parameters.keywords),
```
[the remark this answers is not in the document]
Comment (baiqiushi) [reply examples/twittermap/web/public/javascripts/sidebar/controllers.js]: keywords is an array of String, only `=` assignment is shallow copy, then later check will always will true.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 5]
```diff
@@ -1,4 +1,106 @@
...
+      timeInterval: {
+        start: new Date(cloudberry.parameters.timeInterval.start.getTime()),
```
[the remark this answers is not in the document]
Comment (baiqiushi) [reply examples/twittermap/web/public/javascripts/sidebar/controllers.js]: The same as array, object in Javascript is also shallow copy.

**Pull request merged**